metaflow_extensions/torchrun/plugins/__init__.py

```diff
@@ -46,75 +95,21 @@ def _torchrun_process(self, ubf_context):
-            torchrun_args = {
-                "rdzv-id": "123",
-                "rdzv_endpoint": "%s:%s" % (main_addr, main_port),
-                "nnodes": num_nodes, 
-                "master_addr": main_addr,
-                "master_port": main_port, # batch doesn't matter. k8s depends on port opened in @kubernetes jobset. 
-                "node_rank": node_index,
-                "nproc_per_node": 1, # TODO: set to N visible devices on node
-                "rdzv-backend": "c10d",
-            }
-
-        # apply user overrides in @torchrun_parallel decorator.
-        torchrun_args.update(self.attributes['torchrun_args'])
-
-        # setup torchrun part of command
-        cmd = ["torchrun"]
-        for arg, val in torchrun_args.items():
-            cmd.extend(["--%s" % arg, str(val)])
-
-        # setup torch user script part of command
-        cmd.append(self.attributes['entrypoint'])
-        if isinstance(self.attributes['entrypoint_args'], str):
-            cmd.append(self.attributes['entrypoint_args'])
-        elif isinstance(self.attributes['entrypoint_args'], dict):
-            for arg, val in self.attributes['entrypoint_args'].items():
-                cmd.extend(["--%s" % arg, str(val)])
-        else:
-            raise TypeError("entrypoint_args must be a dict or str")
-
-        self.cmd = " ".join(cmd)
-        subprocess.run(cmd, check=True)
-
-    def task_decorate(
-        self, step_func, flow, graph, retry_count, max_user_code_retries, ubf_context
-    ):
-
-        def _torchrun_with_step_func_postprocessing():
-            self._ensure_torch_installed()
-            self._torchrun_process(ubf_context)
-            setattr(flow, "command", " ".join(self.cmd))
-            step_func()
-        if (
-            ubf_context == UBF_CONTROL
-            and os.environ.get("METAFLOW_RUNTIME_ENVIRONMENT", "local") == "local"
-        ):
-            from functools import partial
-            env_to_use = getattr(self.environment, "base_env", self.environment)
-            return partial(
-                _local_multinode_control_task_step_func,
-                flow,
-                env_to_use,
-                _torchrun_with_step_func_postprocessing,
-                retry_count,
-            )
-        else:
-            return _torchrun_with_step_func_postprocessing
-
-    def _ensure_torch_installed(self):
-        try:
-            import torch
-        except ImportError:
-            print("PyTorch is not installed. Installing the latest version of the torch package from PyPi.")
-            subprocess.run([sys.executable, "-m", "pip", "install", "-U", "torch"], check=True, stdout=subprocess.DEVNULL, stderr=subprocess.STDOUT)
-            import torch
-
+        torch_executor = TorchRunExecutor(
+            pathspec=current.pathspec,
+            main_addr=main_addr,
+            main_port=main_port,
+            num_nodes=num_nodes,
+            node_index=node_index, 
+        )
```

Comment:
You can give this thing a logger, not needed immediately but can be helpful in the near future for debugging internally.